session/txn.go

```diff
@@ -53,6 +53,16 @@ type TxnState struct {
 	doNotCommit error
 }
 
+var _ kv.StoreContract = &TxnState{}
+
+func (st *TxnState) SetContract(key kv.Key, contract kv.ContractType) {
+	if st.Transaction != nil {
+		if raw, ok := st.Transaction.(kv.StoreContract); ok {
```

Comment:
It is weird to use a global variable here.

Reply:
It's not global variable, it's type assert...
I think put the method in the `Transaction` interface is weird, so I define a new interface, and assert that the st.Transaction implement that interface.